enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,105 @@
+As OpenShift V4 is composed of many operators, we need to give operators abiltity to run some checks to findout if the new version is compatible with the existing version and use the result to improve the successrate of upgrades.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before the actual upgrade.
+It would help operators to findout if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered to run these preflight checks without doing an upgrade.
+
+
+### Goals
+
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
+
+### Non-Goals
+
+* This enhancement will not add code required for preflight checks for individual operators.
+* Operators will have to figurout the actual code they would like to run as part of preflight checks.
+* The preflight checks should not be used as a migration mechanism.
+
+## Proposal
+
+Operators need to provide a different bin path for the preflight checks i.e. /bin/preflight.
+The binpath need to exist inside of the operator image.
+Cluster version operator (CVO) will be running a single job per operator for the preflight checks.
+The binpath /bin/preflight must be executable and it can be symbolic link to the operator binary or a script or It can be non existant.
+CVO will not fail if the path does not exist in an operator image but it would assume that preflight checks are not implemented for the respective operator.
+Also there would be a possible way to skip the preflight checks if someone wants to skip it explictly.
+
+### User Stories
+
+#### Story 1
+
+Some customers would prefer to disable automatic credential provisioning by Cloud credential operator (CCO) and instead manage it themselves.
+However situations can arise during an upgrade where credentials have changed and the upgrade will be block and go into a failed state, with little visibility as to why or how it can be fixed.
+With the help of preflight checks CCO can check if they can safely upgrade to a given release before doing the actual upgrade.
+Communicate that an upgrade is not possible as permissions have changed and point administrator with right pointers to fix the issue.
+This would allow users to avoid broken upgrades when credentials have changed and provides them right information to move forward.
+
+### Implementation Details
+
+CVO will run the preflight check from the new operator image to which the curent operator image is trying to upgrade.
+Because the current version of operator does not have knowledge of what new code changes the new version bringing with it.
+So its the new version's responsibility to figureout if the current image can update safely to the new image.
```

Comment:
"figure out"

What namespace will the job run and with what RBAC permissions? I would guess same namespace as the operator and the same service account?

How will preflights report back success or failure, and the detailed info for administrators?

Reply:
Running the *new* image is interesting. In our use case we'll be looking to compare current CredentialsRequests to those in the incoming image to see if anything has changed.

Do you have any ideas yet on what the contract is for information to be provided to the preflight check job, presumably as env vars? I think we'll likely want the incoming release image pullspec and version.

Reply:
> What namespace will the job run and with what RBAC permissions? I would guess same namespace as the operator and the same service account?

yes, you are right. Will add this to the enhancement. 

> How will preflights report back success or failure, and the detailed info for administrators?

This is still an open question.  The ideal would be a JSON object which would be an new API. Or it can just return error/warn/info messages.

Reply:
> ... I think we'll likely want the incoming release image pullspec and version.

Somewhere in this enhancement we should point out that we will require signature verification to pass on the target release before we launch preflights, to guard against malicious images whose preflights attempt to exfiltrate cluster secrets or other sensitive data.  I'm not sure if/how we want to provision for overriding that signature check without overriding preflight results entirely ("I don't care if the target release is signed, but I do care about whether its preflights pass" may not be a thing, and I'm personally fine not supporting it).  Anyhow, once that signature verification is run, we have a trusted version string for the target release, so I'm fine passing that along in the `preflight` (or whatever) calls.

I think:

1. CVO pulls the target release image, extracts referenced images, and runs preflights on them (or a subset of them).
2. Operator-specific preflight gets the target release image pullspec, extracts manifests like CredentialsRequests, and compares them to the in-cluster state.

is... messy.  Unless we pull through our local registry, it could be a bunch of Quay hits.  And it means operators will be vendoring the CVO's "load manifests from `$PULLSPEC`" logic, which is currently only shared between `oc` and the CVO.  Also, the CVO currently uses a job to download and extract the target release image into a shared volume, and if operators follow suite that's download Jobs launched by preflight Jobs and it gets complicated.  If we run these jobs in the `openshift-cluster-version` namespace (also discussed [here][1]), then maybe we could mount the shared volume with extracted manifests into the preflight pods, to save them the trouble of extracting the manifests themselves.  Although I guess that just removes the requirement for preflight-specific target-release download Jobs; they'd still need to vendor or replicate the CVO's "load manifests from a disk" logic.

[1]: https://github.com/openshift/enhancements/pull/363#discussion_r436995128

Reply:
Resolved in https://github.com/openshift/enhancements/commit/4cadad51eb7b2e999146ae9e8af76e57f9c90556 and also in later commits